README.md
# auth-middleware

[![deno land](http://img.shields.io/badge/available%20on-deno.land/x-lightgrey.svg?logo=deno)](https://deno.land/x/auth-middleware)
[![deno doc](https://doc.deno.land/badge.svg)](https://doc.deno.land/https/deno.land/x/auth_middleware/mod.ts)
[![GitHub release (latest by date)](https://img.shields.io/github/v/release/httpland/auth-middleware)](https://github.com/httpland/auth-middleware/releases)
[![codecov](https://codecov.io/github/httpland/auth-middleware/branch/main/graph/badge.svg)](https://codecov.io/gh/httpland/auth-middleware)
[![GitHub](https://img.shields.io/github/license/httpland/auth-middleware)](https://github.com/httpland/auth-middleware/blob/main/LICENSE)

[![test](https://github.com/httpland/auth-middleware/actions/workflows/test.yaml/badge.svg)](https://github.com/httpland/auth-middleware/actions/workflows/test.yaml)
[![NPM](https://nodei.co/npm/@httpland/auth-middleware.png?mini=true)](https://nodei.co/npm/@httpland/auth-middleware/)

HTTP authentication middleware.

Compliant with
[RFC 9110, 11 HTTP Authentication](https://www.rfc-editor.org/rfc/rfc9110.html#name-http-authentication).

## Middleware

For a definition of Universal HTTP middleware, see the
[http-middleware](https://github.com/httpland/http-middleware) project.

## Usage

You specify the Authenticate scheme and provide the authentication function for
token.

```ts
import {
  auth,
  Authentication,
  type Handler,
} from "https://deno.land/x/auth_middleware@$VERSION/mod.ts";

declare const authentication: Authentication;
declare const request: Request;
declare const handler: Handler;

const middleware = auth(authentication);
const response = await middleware(request, handler);
```

## Basic authentication

Provides ready-to-use Authorization for Basic Authentication.

Compliant with
[RFC 7617, The 'Basic' HTTP Authentication Scheme](https://www.rfc-editor.org/rfc/rfc7617).

```ts
import {
  auth,
  Basic,
  type Handler,
  type User,
} from "https://deno.land/x/auth_middleware@$VERSION/mod.ts";
import { assert } from "https://deno.land/std/testing/asserts.ts";
import { assertSpyCalls, spy } from "https://deno.land/std/testing/mock.ts";

declare const admin: User;
declare const request: Request;
declare const handler: Handler;

const middleware = auth(
  new Basic(({ username, password }) => {
    const userResult = Basic.timingSafeEqual(username, admin.username);
    const passResult = Basic.timingSafeEqual(password, admin.password);

    return userResult && passResult;
  }),
);
const spiedHandler = spy(handler);
const response = await middleware(request, spiedHandler);

assertSpyCalls(spiedHandler, 0);
assert(response.headers.has("www-authenticate"));
```

yield:

```http
WWW-Authenticate: Basic realm="Secure area"
```

## License

Copyright © 2023-present [httpland](https://github.com/httpland).

Released under the [MIT](./LICENSE) license


CHANGELOG.md
# [1.0.0-beta.3](https://github.com/httpland/auth-middleware/compare/1.0.0-beta.2...1.0.0-beta.3) (2023-05-03)


### Features

* **bearer:** add bearer authentication ([1c7b498](https://github.com/httpland/auth-middleware/commit/1c7b498205842596495ebcf6da03615ed9ed58ec))
* **types:** change Authentication#authenticate interface ([3549555](https://github.com/httpland/auth-middleware/commit/354955578a01f8eaee41bf4d32d823988e50cbdf))

# [1.0.0-beta.2](https://github.com/httpland/auth-middleware/compare/1.0.0-beta.1...1.0.0-beta.2) (2023-05-02)


### Features

* **basic:** provide timing safe equal ([1c030cd](https://github.com/httpland/auth-middleware/commit/1c030cd6e4f4ca3dc4c033958af3b0d4bbd4ddc2))
* **deps:** update deps version ([46d463d](https://github.com/httpland/auth-middleware/commit/46d463d57df9899b47727b6266dca423afe38fa5))
* **types:** change authentication interface ([6ceba10](https://github.com/httpland/auth-middleware/commit/6ceba104e1f693b442039f238209759053452f52))

# 1.0.0-beta.1 (2023-04-27)


### Bug Fixes

* **digest:** fix to wrong parameter representation ([7e5c0d2](https://github.com/httpland/auth-middleware/commit/7e5c0d2615a679143fcdf28c57afcd66d6cd4589))


### Features

* **basic:** change user interface ([0c7ff1e](https://github.com/httpland/auth-middleware/commit/0c7ff1eb103aba0744aef3c1ea69e6e1134942af))
* **digest:** export related types ([1191d14](https://github.com/httpland/auth-middleware/commit/1191d14504f355e0224b2240b70df5de8c9a7c9a))
* **digest:** support `domain` parameter ([87cf97c](https://github.com/httpland/auth-middleware/commit/87cf97ce2cb9fa269fb94842f2983f0f73e58423))
* **digest:** support `opaque` field ([53cb8d4](https://github.com/httpland/auth-middleware/commit/53cb8d43f8cf496e79099519c1e44668d73478e3))
* **middleware:** add auth middleware factory ([80ee979](https://github.com/httpland/auth-middleware/commit/80ee979bafd27285650123a1670842d414920cc9))
* **mod:** export utilities ([ca2fac9](https://github.com/httpland/auth-middleware/commit/ca2fac96def180a5921ccb15f9f232886d0af38b))
* **schemes:** add basic authentication scheme ([94d18f9](https://github.com/httpland/auth-middleware/commit/94d18f9ccd8ada2468e027e4ae906f881fe94e82))
* **schemes:** add minimum Digest scheme ([3490f44](https://github.com/httpland/auth-middleware/commit/3490f440ad5374db98f79a7dd053eed6a401410a))
* **schemes:** change digest scheme interface ([4f5004a](https://github.com/httpland/auth-middleware/commit/4f5004a470c67fa20af9ad84ccf6e2904a223b09))
* **schemes:** change interface to accept authorizer ([87148a5](https://github.com/httpland/auth-middleware/commit/87148a5f51000176ec92b68c172b9d2a43433ff7))
* **schemes:** change interface to accept promise ([390f82d](https://github.com/httpland/auth-middleware/commit/390f82d781cb07497f10fb79d25e19db65f24467))
* **schemes:** digest scheme support SHA-256 ([695cc29](https://github.com/httpland/auth-middleware/commit/695cc29b030e924f21c5d3f530c92fbc33964866))
* **schemes:** support nonce and charset field ([97eaddd](https://github.com/httpland/auth-middleware/commit/97eaddd5168fdd8f351f3915ebe755360ad17645))
* **schemes:** support SHA-512/256 ([4714d13](https://github.com/httpland/auth-middleware/commit/4714d131009898e864310c71668cbf600a1ac24b))


schemes/digest.ts
// Copyright 2023-latest the httpland authors. All rights reserved. MIT license.
// This module is browser compatible.

import {
  AuthParams,
  isObject,
  isString,
  isUndefined,
  stringifyChallenge,
  toHashString,
} from "../deps.ts";
import {
  badRequest,
  concat,
  err,
  ok,
  omitBy,
  type Quoted,
  quoted,
  timingSafeEqual,
  unauthorized,
} from "../utils.ts";
import { Char, DEFAULT_REALM } from "../constants.ts";
import type {
  Authentication,
  AuthenticationResult,
  AuthParamsContext,
  Realm,
  User,
} from "../types.ts";
import { Crypto } from "./_wasm/mod.ts";

const enum Algorithm {
  MD5 = "MD5",
  MD5sess = `${Algorithm.MD5}-sess`,
  SHA_256 = "SHA-256",
  SHA_256sess = `${Algorithm.SHA_256}-sess`,
  SHA_512_256 = "SHA-512-256",
  SHA_512_256sess = `${Algorithm.SHA_512_256}-sess`,
}

const algorithms = [
  Algorithm.MD5,
  Algorithm.MD5sess,
  Algorithm.SHA_256,
  Algorithm.SHA_256sess,
  Algorithm.SHA_512_256,
  Algorithm.SHA_512_256sess,
];

/** Digest algorithm. */
export type DigestAlgorithm = `${Algorithm}`;

/** Digest options. */
export interface DigestOptions {
  /** Algorithm. */
  readonly algorithm?: DigestAlgorithm;

  /** Realm.
   * @default "Secure area"
   */
  readonly realm?: string;

  /** Calculate nonce. */
  readonly nonce?: () => string | Promise<string>;

  /** A string of data, specified by the server.
   * It is recommended that this string be Base64 or hexadecimal data.
   */
  readonly opaque?: string;

  /** Protection space specified with URI. */
  readonly domain?: string[];

  /** Encoding scheme. */
  readonly charset?: "UTF-8";

  // TODO:(miyauci) support features
  // readonly qop?: readonly QOP[];
  // readonly userhash?: boolean;
}

/** User selection API. */
export interface SelectUser {
  (username: string): User | void | Promise<User | void>;
}

/** HTTP Digest authentication. */
export class Digest implements Authentication {
  scheme = "Digest";

  #staticOptions: AuthParams;
  #algorithm: DigestAlgorithm;
  #nonce: () => string | Promise<string>;
  #selectUser: SelectUser;

  constructor(
    selectUser: SelectUser,
    options: DigestOptions = {},
  ) {
    const {
      realm = DEFAULT_REALM,
      algorithm,
      charset,
      nonce = calculateNonce,
      opaque,
      domain,
    } = options;
    const params: Omit<DigestRequestParams, "nonce"> = {
      charset,
      realm: quoted(realm),
      qop: `"auth"`,
      algorithm,
      opaque: isString(opaque) ? quoted(opaque) : undefined,
      domain: domain && quoted(domain.join(" ")),
    };

    this.#staticOptions = omitBy(params, isUndefined);
    this.#algorithm = algorithm ?? Algorithm.MD5;
    this.#nonce = nonce;
    this.#selectUser = selectUser;
  }

  async authenticate(
    context:
      & { readonly request: Pick<Request, "method"> }
      & Pick<AuthParamsContext, "params">,
  ): Promise<AuthenticationResult> {
    const { request, params } = context;

    if (!isObject(params)) {
      return err(badRequest(`Authorization challenge should be valid`));
    }

    if (!isDigestResponseParams(params)) {
      return err(badRequest(`auth-param should be valid`));
    }

    const {
      nc,
      cnonce,
      qop,
      nonce,
      uri,
      response,
      realm,
      algorithm = Algorithm.MD5,
    } = params;

    if (!isDigestAlgorithm(algorithm)) {
      return err(
        badRequest(
          `auth-param of "algorithm" should be one of ${
            algorithms.map(quoted).join(", ")
          }`,
        ),
      );
    }

    const username = unq(params.username);
    const maybeUser = await this.#selectUser(username);

    if (!maybeUser || !timingSafeEqual(username, maybeUser.username)) {
      return err(unauthorized(await this.challenge()));
    }

    const res = calculateResponse({
      cnonce: unq(cnonce),
      method: request.method,
      nc: unq(nc),
      nonce: unq(nonce),
      qop,
      realm: unq(realm),
      secret: maybeUser.password,
      uri: unq(uri),
      username: maybeUser.username,
      algorithm,
    });

    const result = timingSafeEqual(unq(response), res);

    if (!result) return err(unauthorized(await this.challenge()));

    return ok();
  }

  async challenge(): Promise<string> {
    const nonce = await this.#nonce();
    const challenge = stringifyChallenge({
      authScheme: this.scheme,
      params: { ...this.#staticOptions, nonce: quoted(nonce) },
    });

    return challenge;
  }
}

const Supported = {
  [Algorithm.MD5]: md5,
  [Algorithm.SHA_256]: sha256,
  [Algorithm.SHA_512_256]: sha512_256,
};

/** Whether the input is {@link DigestAlgorithm} or not. */
export function isDigestAlgorithm(input: string): input is DigestAlgorithm {
  switch (input) {
    case Algorithm.MD5:
    case Algorithm.MD5sess:
    case Algorithm.SHA_256sess:
    case Algorithm.SHA_256:
    case Algorithm.SHA_512_256:
    case Algorithm.SHA_512_256sess: {
      return true;
    }

    default: {
      return false;
    }
  }
}

/** Un-quoted string. */
export function unq(input: string): string {
  if (input.startsWith(Char.DQuote) && input.endsWith(Char.DQuote)) {
    return input.slice(1, -1);
  }

  return input;
}

function calculateNonce(): string {
  return btoa(crypto.randomUUID());
}

export function md5(input: string): string {
  return toHashString(
    Crypto.subtle.digestSync(Algorithm.MD5, new TextEncoder().encode(input)),
  );
}

export function sha512_256(input: string): string {
  return toHashString(
    Crypto.subtle.digestSync(
      Algorithm.SHA_512_256,
      new TextEncoder().encode(input),
    ),
  );
}

export function calculateResponse(context: {
  username: string;
  nonce: string;
  secret: string;
  method: string;
  uri: string;
  realm: string;
  nc: string;
  cnonce: string;
  qop: string;
  algorithm: DigestAlgorithm;
}): string {
  const {
    method,
    uri,
    username,
    realm,
    secret,
    nonce,
    nc,
    cnonce,
    qop,
    algorithm,
  } = context;
  const sess = algorithm.endsWith("sess");
  const H = Supported[normalizeAlgorithm(algorithm as Algorithm)];
  const _A1 = concat(username, ":", realm, ":", secret);
  const A1 = sess ? concat(H(_A1), ":", nonce, ":", cnonce) : _A1;
  const A2 = `${method}:${uri}`;

  /** Calculate key for digest.
   * @see [RFC 7616, 3.3.  The WWW-Authenticate Response Header Field](https://datatracker.ietf.org/doc/html/rfc7616#section-3.3)
   */
  function KD(secret: string, data: string): string {
    return H(concat(secret, ":", data));
  }

  const response = KD(H(A1), `${nonce}:${nc}:${cnonce}:${qop}:${H(A2)}`);

  return response;
}

export function sha256(input: string): string {
  return toHashString(
    Crypto.subtle.digestSync(
      Algorithm.SHA_256,
      new TextEncoder().encode(input),
    ),
  );
}

function normalizeAlgorithm(
  algorithm: Algorithm,
): Algorithm.MD5 | Algorithm.SHA_256 | Algorithm.SHA_512_256 {
  switch (algorithm) {
    case Algorithm.MD5sess: {
      return Algorithm.MD5;
    }
    case Algorithm.SHA_256sess: {
      return Algorithm.SHA_256;
    }

    case Algorithm.SHA_512_256sess: {
      return Algorithm.SHA_512_256;
    }

    default: {
      return algorithm;
    }
  }
}

interface DigestParams extends Realm {
  readonly nonce: Quoted;
  readonly opaque?: Quoted;
}

interface DigestRequestParams extends DigestParams {
  readonly algorithm?: `${Algorithm}`;
  readonly charset?: "UTF-8";
  readonly domain?: Quoted;
  readonly qop: Quoted;
}

export interface DigestResponseParams extends DigestParams {
  readonly response: Quoted;
  readonly username: Quoted;
  readonly uri: Quoted;
  readonly cnonce: Quoted;
  readonly nc: string;
  readonly qop: string;
}

function isDigestResponseParams(
  // deno-lint-ignore no-explicit-any
  authParams: Record<string, any>,
): authParams is DigestResponseParams {
  return Object
    .values(DigestParamKey)
    .every((key) => key in authParams);
}

enum DigestParamKey {
  Response = "response",
  Username = "username",
  Realm = "realm",
  URI = "uri",
  Cnonce = "cnonce",
  Nc = "nc",
  Nonce = "nonce",
  QOP = "qop",
}
